File: backend/app/api/v1/expose.py
"""
Exposé API Endpoints
"""
from typing import List
from fastapi import APIRouter, Depends, HTTPException, status
from fastapi.responses import Response

from app.api.deps import require_read_scope, require_write_scope, get_tenant_id
from app.core.security import TokenData
from app.core.errors import NotFoundError
from app.schemas.expose import (
    ExposeGenerateRequest, ExposeGenerateResponse,
    ExposeSaveRequest, ExposeVersionResponse,
    ExposeListResponse, ExposePDFRequest, ExposePDFResponse
)
from app.services.expose_service import ExposeService
from app.services.pdf_generator_service import PDFGeneratorService
from app.db.models import ExposeVersion

router = APIRouter()


@router.post("/{property_id}/expose/generate", response_model=ExposeGenerateResponse)
async def generate_expose(
    property_id: str,
    request: ExposeGenerateRequest,
    current_user: TokenData = Depends(require_write_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Generate exposé using LLM"""
    
    expose_service = ExposeService(tenant_id)
    
    try:
        version = await expose_service.generate_expose(
            property_id=property_id,
            audience=request.audience,
            tone=request.tone,
            language=request.language,
            length=request.length,
            keywords=request.keywords,
            user_id=current_user.user_id
        )
        
        return ExposeGenerateResponse(
            version=ExposeVersionResponse(
                id=str(version.id),
                title=version.title,
                content=version.content,
                audience=version.audience,
                tone=version.tone,
                language=version.language,
                length=version.length,
                keywords=version.keywords,
                status=version.status,
                version_number=version.version_number,
                created_at=version.created_at.isoformat(),
                updated_at=version.updated_at.isoformat(),
                created_by=str(version.created_by.id)
            ),
            generated_at=version.created_at.isoformat()
        )
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate exposé: {str(e)}"
        )


@router.get("/{property_id}/expose/versions", response_model=ExposeListResponse)
async def get_expose_versions(
    property_id: str,
    current_user: TokenData = Depends(require_read_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Get all exposé versions for a property"""
    
    expose_service = ExposeService(tenant_id)
    
    try:
        versions = await expose_service.get_expose_versions(property_id)
        
        version_responses = [
            ExposeVersionResponse(
                id=str(version.id),
                title=version.title,
                content=version.content,
                audience=version.audience,
                tone=version.tone,
                language=version.language,
                length=version.length,
                keywords=version.keywords,
                status=version.status,
                version_number=version.version_number,
                created_at=version.created_at.isoformat(),
                updated_at=version.updated_at.isoformat(),
                created_by=str(version.created_by.id)
            )
            for version in versions
        ]
        
        return ExposeListResponse(
            versions=version_responses,
            total=len(version_responses)
        )
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to get exposé versions: {str(e)}"
        )


@router.post("/{property_id}/expose/save", response_model=ExposeVersionResponse)
async def save_expose_version(
    property_id: str,
    request: ExposeSaveRequest,
    current_user: TokenData = Depends(require_write_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Save exposé version"""
    
    expose_service = ExposeService(tenant_id)
    
    try:
        version = await expose_service.save_expose_version(
            property_id=property_id,
            title=request.title,
            content=request.content,
            audience=request.audience,
            tone=request.tone,
            language=request.language,
            length=request.length,
            keywords=request.keywords,
            user_id=current_user.user_id
        )
        
        return ExposeVersionResponse(
            id=str(version.id),
            title=version.title,
            content=version.content,
            audience=version.audience,
            tone=version.tone,
            language=version.language,
            length=version.length,
            keywords=version.keywords,
            status=version.status,
            version_number=version.version_number,
            created_at=version.created_at.isoformat(),
            updated_at=version.updated_at.isoformat(),
            created_by=str(version.created_by.id)
        )
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to save exposé: {str(e)}"
        )


@router.delete("/{property_id}/expose/{version_id}")
async def delete_expose_version(
    property_id: str,
    version_id: str,
    current_user: TokenData = Depends(require_write_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Delete exposé version"""
    
    expose_service = ExposeService(tenant_id)
    
    try:
        await expose_service.delete_expose_version(version_id, current_user.user_id)
        return {"message": "Exposé version deleted successfully"}
        
    except NotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exposé version not found"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to delete exposé: {str(e)}"
        )


@router.post("/{property_id}/expose/{version_id}/publish", response_model=ExposeVersionResponse)
async def publish_expose_version(
    property_id: str,
    version_id: str,
    current_user: TokenData = Depends(require_write_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Publish exposé version"""
    
    expose_service = ExposeService(tenant_id)
    
    try:
        version = await expose_service.publish_expose_version(version_id, current_user.user_id)
        
        return ExposeVersionResponse(
            id=str(version.id),
            title=version.title,
            content=version.content,
            audience=version.audience,
            tone=version.tone,
            language=version.language,
            length=version.length,
            keywords=version.keywords,
            status=version.status,
            version_number=version.version_number,
            created_at=version.created_at.isoformat(),
            updated_at=version.updated_at.isoformat(),
            created_by=str(version.created_by.id)
        )
        
    except NotFoundError:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exposé version not found"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to publish exposé: {str(e)}"
        )


@router.post("/{property_id}/expose/pdf", response_model=ExposePDFResponse)
async def generate_expose_pdf(
    property_id: str,
    request: ExposePDFRequest,
    current_user: TokenData = Depends(require_write_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Generate exposé PDF"""
    
    # Get exposé version
    from asgiref.sync import sync_to_async
    
    @sync_to_async
    def get_expose_version():
        try:
            return ExposeVersion.objects.get(id=request.version_id)
        except ExposeVersion.DoesNotExist:
            return None
    
    version = await get_expose_version()
    if not version:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exposé version not found"
        )
    
    # Generate PDF using the same service as energy certificate
    pdf_service = PDFGeneratorService()
    
    # Prepare property data
    property_data = {
        'title': version.property.title,
        'address': {
            'street': getattr(version.property.address, 'street', '') if version.property.address else '',
            'zip_code': getattr(version.property.address, 'zip_code', '') if version.property.address else '',
            'city': getattr(version.property.address, 'city', '') if version.property.address else '',
        }
    }
    
    # Prepare exposé data
    expose_data = {
        'title': version.title,
        'content': version.content,
        'audience': version.audience,
        'tone': version.tone,
        'language': version.language,
        'length': version.length,
        'keywords': version.keywords
    }
    
    # Company data
    company_data = {
        'company_name': 'ImmoNow',
        'address': 'Musterstraße 123, 12345 Musterstadt',
        'phone': '+49 123 456789',
        'email': 'info@immonow.de'
    }
    
    try:
        # Generate PDF (using a modified method for exposé)
        pdf_bytes = pdf_service.generate_energy_certificate_pdf(
            property_data=property_data,
            energy_data=expose_data,  # Reusing the same method structure
            company_data=company_data,
            logo_path=None,
            language=version.language
        )
        
        # Save PDF (placeholder)
        filename = f"Expose_{version.property.title}_{version.version_number}_{version.created_at.strftime('%Y%m%d')}.pdf"
        pdf_url = f"/media/exposes/{filename}"
        
        return ExposePDFResponse(
            pdf_url=pdf_url,
            filename=filename,
            generated_at=version.created_at.isoformat()
        )
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate PDF: {str(e)}"
        )


@router.get("/{property_id}/expose/{version_id}/download")
async def download_expose_pdf(
    property_id: str,
    version_id: str,
    current_user: TokenData = Depends(require_read_scope),
    tenant_id: str = Depends(get_tenant_id)
):
    """Download exposé PDF"""
    
    # Get exposé version
    from asgiref.sync import sync_to_async
    
    @sync_to_async
    def get_expose_version():
        try:
            return ExposeVersion.objects.get(id=version_id)
        except ExposeVersion.DoesNotExist:
            return None
    
    version = await get_expose_version()
    if not version:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Exposé version not found"
        )
    
    # Generate PDF
    pdf_service = PDFGeneratorService()
    
    property_data = {
        'title': version.property.title,
        'address': {
            'street': getattr(version.property.address, 'street', '') if version.property.address else '',
            'zip_code': getattr(version.property.address, 'zip_code', '') if version.property.address else '',
            'city': getattr(version.property.address, 'city', '') if version.property.address else '',
        }
    }
    
    expose_data = {
        'title': version.title,
        'content': version.content,
        'audience': version.audience,
        'tone': version.tone,
        'language': version.language,
        'length': version.length,
        'keywords': version.keywords
    }
    
    company_data = {
        'company_name': 'ImmoNow',
        'address': 'Musterstraße 123, 12345 Musterstadt',
        'phone': '+49 123 456789',
        'email': 'info@immonow.de'
    }
    
    try:
        pdf_bytes = pdf_service.generate_energy_certificate_pdf(
            property_data=property_data,
            energy_data=expose_data,
            company_data=company_data,
            logo_path=None,
            language=version.language
        )
        
        filename = f"Expose_{version.property.title}_v{version.version_number}.pdf"
        
        return Response(
            content=pdf_bytes,
            media_type="application/pdf",
            headers={"Content-Disposition": f"attachment; filename={filename}"}
        )
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to generate PDF: {str(e)}"
        )

File: frontend/src/services/expose.ts
/**
 * Exposé Service
 * API calls for exposé management
 */

import { apiClient } from '../api/config';

export interface ExposeVersionData {
  id: string;
  title: string;
  content: string;
  audience: string;
  tone: string;
  language: string;
  length: string;
  keywords: string[];
  status: string;
  version_number: number;
  created_at: string;
  updated_at: string;
  created_by: string;
}

export interface ExposeGenerateRequest {
  audience: string;
  tone: string;
  language?: string;
  length?: string;
  keywords?: string[];
}

export interface ExposeSaveRequest {
  title: string;
  content: string;
  audience: string;
  tone: string;
  language?: string;
  length?: string;
  keywords?: string[];
}

export interface ExposeGenerateResponse {
  version: ExposeVersionData;
  generated_at: string;
}

export interface ExposeListResponse {
  versions: ExposeVersionData[];
  total: number;
}

export interface ExposePDFRequest {
  version_id: string;
  include_logo?: boolean;
  template?: string;
}

export interface ExposePDFResponse {
  pdf_url: string;
  filename: string;
  generated_at: string;
  version_id?: string;
}

export class ExposeService {
  /**
   * Generate exposé using LLM
   */
  static async generateExpose(
    propertyId: string,
    request: ExposeGenerateRequest
  ): Promise<ExposeGenerateResponse> {
    try {
      const response = await apiClient.post<ExposeGenerateResponse>(
        `/api/v1/properties/${propertyId}/expose/generate`,
        {
          audience: request.audience,
          tone: request.tone,
          language: request.language || 'de',
          length: request.length || 'standard',
          keywords: request.keywords || []
        }
      );
      return response;
    } catch (error: any) {
      console.error('Error generating exposé:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Generieren des Exposés');
    }
  }

  /**
   * Get all exposé versions for a property
   */
  static async getExposeVersions(propertyId: string): Promise<ExposeListResponse> {
    try {
      const response = await apiClient.get<ExposeListResponse>(
        `/api/v1/properties/${propertyId}/expose/versions`
      );
      return response;
    } catch (error: any) {
      console.error('Error fetching exposé versions:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Laden der Exposé-Versionen');
    }
  }

  /**
   * Save exposé version
   */
  static async saveExposeVersion(
    propertyId: string,
    request: ExposeSaveRequest
  ): Promise<ExposeVersionData> {
    try {
      const response = await apiClient.post<ExposeVersionData>(
        `/api/v1/properties/${propertyId}/expose/save`,
        {
          title: request.title,
          content: request.content,
          audience: request.audience,
          tone: request.tone,
          language: request.language || 'de',
          length: request.length || 'standard',
          keywords: request.keywords || []
        }
      );
      return response;
    } catch (error: any) {
      console.error('Error saving exposé:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Speichern des Exposés');
    }
  }

  /**
   * Delete exposé version
   */
  static async deleteExposeVersion(propertyId: string, versionId: string): Promise<void> {
    try {
      await apiClient.delete(
        `/api/v1/properties/${propertyId}/expose/${versionId}`
      );
    } catch (error: any) {
      console.error('Error deleting exposé version:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Löschen der Exposé-Version');
    }
  }

  /**
   * Publish exposé version
   */
  static async publishExposeVersion(propertyId: string, versionId: string): Promise<ExposeVersionData> {
    try {
      const response = await apiClient.post<ExposeVersionData>(
        `/api/v1/properties/${propertyId}/expose/${versionId}/publish`
      );
      return response;
    } catch (error: any) {
      console.error('Error publishing exposé:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Veröffentlichen des Exposés');
    }
  }

  /**
   * Generate exposé PDF
   */
  static async generatePDF(
    propertyId: string,
    request: ExposePDFRequest
  ): Promise<ExposePDFResponse> {
    try {
      const response = await apiClient.post<ExposePDFResponse>(
        `/api/v1/properties/${propertyId}/expose/pdf`,
        {
          version_id: request.version_id,
          include_logo: request.include_logo !== false,
          template: request.template || 'standard'
        }
      );
      return response;
    } catch (error: any) {
      console.error('Error generating PDF:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Generieren des PDFs');
    }
  }

  /**
   * Download exposé PDF
   */
  static async downloadPDF(propertyId: string, versionId: string): Promise<Blob> {
    try {
      const response = await apiClient.get(
        `/api/v1/properties/${propertyId}/expose/${versionId}/download`,
        {
          responseType: 'blob'
        }
      );
      return response as unknown as Blob;
    } catch (error: any) {
      console.error('Error downloading PDF:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Herunterladen des PDFs');
    }
  }

  /**
   * Download PDF and trigger browser download
   */
  static async downloadPDFFile(propertyId: string, versionId: string, filename?: string): Promise<void> {
    try {
      const blob = await this.downloadPDF(propertyId, versionId);

      // Create download link
      const url = window.URL.createObjectURL(blob);
      const link = document.createElement('a');
      link.href = url;
      link.download = filename || `Expose_${propertyId}_${new Date().toISOString().split('T')[0]}.pdf`;

      // Trigger download
      document.body.appendChild(link);
      link.click();
      document.body.removeChild(link);

      // Clean up
      window.URL.revokeObjectURL(url);
    } catch (error: any) {
      console.error('Error downloading PDF file:', error);
      throw new Error(error.response?.data?.detail || 'Fehler beim Herunterladen des PDFs');
    }
  }
}
